lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs

```diff
@@ -0,0 +1,221 @@
+        /// <param name="initialValue"></param>
+        internal MetricTracker(IMetric metric)
+        {
+            Subscribe(metric);
+            _mirror = metric;
+            ChangesSinceLastSink = 0;
+            _keepUpdateHistory = metric.IsImmutable;
+            _records = new ConcurrentQueue<MetricRecord>();
+            _records.Enqueue(CreateMetricRecord(metric));
+        }
+
+        [JsonConstructor]
+        internal MetricTracker(IMetric _mirRor, int changesSinceLastSink, ConcurrentQueue<MetricRecord> _recorDs)
+        {
+            _mirror = _mirRor;
+            _records = new ConcurrentQueue<MetricRecord>(_recorDs);
+            ChangesSinceLastSink = changesSinceLastSink;
+        }
+
+        private MetricTracker(IMetric mirror, ConcurrentQueue<MetricRecord> records, int changes, bool history)
+        {
+            _mirror = mirror;
+            _records = records;
+            ChangesSinceLastSink = changes;
+            _keepUpdateHistory = history;
+        }
+
+        /// <summary>
+        /// Flush records.
+        /// </summary>
+        internal ConcurrentQueue<MetricRecord> FlushChangesSinceLastSink()
+        {
+            ConcurrentQueue<MetricRecord> records = new ConcurrentQueue<MetricRecord>();
+            MetricRecord record;
+            while (_records.TryDequeue(out record))
+            {
+                records.Enqueue(record);
+            }
+            ChangesSinceLastSink = 0;
+            return records;
+        }
+
+        /// <summary>
+        /// When new metric data is received, update the value and records so it reflects the new data.
+        /// </summary>
+        /// <param name="metric">Metric data received.</param>
+        internal MetricTracker UpdateMetric(MetricTracker metric)
+        {
+            _mirror = metric.GetMetric();
+            if (metric.ChangesSinceLastSink > 0)
+            {
+                if (_keepUpdateHistory)
+                {
+                    _records.Concat(metric._records);
+                }
+                else
+                {
+                    // _records = metric.GetMetricRecords().ToList();
+                    _records = metric.GetMetricRecords();
+                }
+            }
+            ChangesSinceLastSink += metric.ChangesSinceLastSink;
+            return this;
+        }
+
+        /// <summary>
+        /// Updates metric value with metric object received.
+        /// </summary>
+        /// <param name="me">New metric.</param>
+        internal void UpdateMetric(IMetric me)
+        {
+            ChangesSinceLastSink++;
+            _mirror = me;
+            UpdateRecords();
+        }
+
+        private void UpdateRecords()
+        {
+            var newRecord = CreateMetricRecord(_mirror);
+            if (_keepUpdateHistory)
+            {
+                _records.Enqueue(newRecord);
+            }
+            else
+            {
+                _records = new ConcurrentQueue<MetricRecord>();
```

Comment:
Talked about this offline. 1) We are allocating a concurrent queue each time we increment. 2) This assignment and enqueueing are not thread safe and could cause some undersirable outcomes. Mandy was looking into creating two different types of trackers to handle this.